### PR TITLE
Don't challenge without GPS lock close to assert

### PR DESCRIPTION
**Pull request by evanmcc**
This is a WIP version to get some feedback on the approach here.

The idea is that we skim GPS info off of our status packet and store it in the `miner_lora` server.  Then we check if it's close enough to our asserted position to see if we're allowed to make any requests.

Main questions here: 
1. Is 75 meters too lenient? Too strict?
2. Some other GPS using PRs we do some additional position adjustment, do we need to do those here?

### CODE REVIEW

**Comment by fvasquez** [comment]
> 1. Is 75 meters too lenient? Too strict?

I think 75 meters may be too lenient. On average, smartphone GPS location accuracy is 30 meters in dense urban areas.

**Comment by evanmcc** [comment]
I guess I'd rather start too lenient and crank it down if too many people are slipping through.  With 30m accuracy, 75m might not be lenient enough.

**Comment by evanmcc** [comment]
commit 83fe6b9 will require merging https://github.com/helium/blockchain-core/pull/445 first to work

commit bc12b29 should be reverted for testing and to make the feature active

**Comment by fasuloj** [comment]
I've left gps trackers in my parked vehicle dashboard and it has veered off a quarter mile in any direction on cargo. Cargo shows it as if if moved around the area even though it is in a stationary position.. I would expect the same drift from my hotspots GPS position.

**Comment by evanmcc** [comment]
Hi @fasuloj!

Rest assured that there will be a long data gathering and analysis period before we deploy this change widely.  The PR as currently written implements the feature in a kind of data gathering mode.  It adds anonymized informatoin to the gossip metadata, so we can figure out how many people would be affected by this.  There will be more official communications in the coming weeks about how the deployment is going to go.

File: src/miner_lora.erl [hunk 1]
```diff
@@ -131,6 +140,36 @@ handle_call({send, Payload, When, Freq, DataRate, Power, IPol}, From, #state{soc
...
+                undefined ->
+                    {error, not_asserted};
+                H3Loc ->
+                    case vincenty:distance(h3:to_geo(H3Loc), State#state.latlong) of
+                        {error, _E} ->
+                            lager:debug("fix error! ~p", [_E]),
+                            {error, bad_calculation};
+                        {ok, Distance} when Distance > ?MAX_WANDER_DIST ->
+                            lager:debug("fix too far! ~p", [Distance]),
+                            {ok, bad_assert, State#state.latlong};
```
Comment (vihu) [inline src/miner_lora.erl:161]: This return breaks the type spec contract

File: src/poc/miner_poc_statem.erl [hunk 2]
```diff
@@ -623,28 +622,40 @@ allow_request(BlockHash, #data{blockchain=Blockchain,
...
+                    true;
+                LastChallenge ->
+                    case (Height - LastChallenge) > POCInterval of
+                        true -> 1 == rand:uniform(trunc(POCInterval / 4));
+                        false -> false
                     end
-            end
+            end,
+        LocationOK = true,
+        %% the below code is tested and working.  we're going to roll
```
Comment (Vagabond) [inline src/poc/miner_poc_statem.erl:641]: You should probably also not report challenges or witnesses if you have this issue as well.

**Comment by evanmcc** [comment]
6763e04 is the new commit to revert to activate this feature.

File: src/poc/miner_poc_statem.erl (hunk 2 above)
Comment (evanmcc) [reply src/poc/miner_poc_statem.erl]: should be remedied by a167a2d

**Pull request merged**